### PR TITLE
HTTP/2 stream removed from map before onStreamClosed called

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
The interface contract of Http2Connection.Listener.onStreamClosed says that the stream will be removed from the active stream map, and not necessarily the stream map. If the channel becomes inactive we may remove from the stream map before calling onStreamClosed.

Modifications:
- Don't remove from the stream map during iteration until after onStreamClosed is called

Result:
Expectations of onStreamClosed interface are not violated

### CODE REVIEW

**Comment by Scottmitch** [comment]
@nmittler @normanmaurer - PTAL.

It is debatable whether the onStreamClosed interface should also include the stream is no longer guaranteed to be in the map, but this PR just ensures the behavior remains consistent.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 1]
```diff
@@ -147,13 +136,12 @@ public void operationComplete(Future<Void> future) throws Exception {
         // paths iterating over the active streams.
         if (activeStreams.allowModifications()) {
             while (itr.hasNext()) {
-                Http2Stream stream = itr.next().value();
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Why this change?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 2]
```diff
@@ -1178,14 +1174,14 @@ public void process() {
             }
         }
 
-        public void deactivate(final DefaultStream stream) {
+        public void deactivate(final DefaultStream stream, final Iterator<?> itr) {
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Can you remind me why this method is public?  I don't think we should be exposing these iterators on public interfaces.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 1 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: So I could add a package private method `close(Iterator<?> itr)` on `DefaultStream`. This `itr` is used to do the removal from `streamMap` if present to prevent any chance of concurrent modification.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 2 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: the class is private. all of the public methods could be made package private, but I'm not sure if it matters much. WDYT?

Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: +1 @nmittler anything else ?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 1 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: oh I see ... nevermind, I'm blind :)

**Comment by nmittler** [comment]
@Scottmitch LGTM

**Comment by Scottmitch** [comment]
Cherry-picked c4fbc06
